### ISSUE TITLE
Make ReachContinuous and ReachDiscrete handle nTimeGridPoints property correctly

### ISSUE THREAD

**Issue by pgagarinov** [SystemAnalysisDpt-CMC-MSU/ellipsoids#12]
Deadline - 30th of October, by that time all changes and tests should be merged into "master" branch
---------------------------------------------------------------------------------------------------------------
1) elltool.reach.ReachContinuous class doesn't seem to take nTimeGrindPoints property into account - in elltool.demo.test.control.anim6 function for instance calling elltool.conf.Properties.setNTimeGridPoints(135); doesn't change a number of time points in returned ellipsoidal tube (100). That is because nTimeGridPoints is not passed into the low-level EllTube builder class.
2) For discrete systems this property seems to be ignored altogether though it should be set based on time limits.

One needs to fix these two problems and write tests in a test case class located inproducts\+elltool\+reach\+test\+mlunit package. The tests should check explicitly that these properties are not ignored for a few different systems (both discrete and continous) in both direct and opposite time, for projections and not projections. One can either create a separate test case class and put the tests in one of the existing test cases.

**Comment by pgagarinov** [comment]
You still need to write tests for both continuous and discrete systems.

**Comment by pgagarinov** [comment]
I do not see any major differences from what you had in Nov - still no tests.

**Comment by ruslan-malsagov** [comment]
Good evening, Peter.
Could you please help me: I made 5 insertions of an additional argument "nTimeGridPoints" in a few functions, so that now it seems that "elltool.demo.test.control.anim6" takes this property into account. Then I commited and pushed my branch and I got test results: 48 errors and 2 failures. Does it mean that I should edit existing tests somehow to get rid of these errors and failures or it means that I did something wrong with my insertions of the argument "nTimeGridPoints"? Or I just need to write my own tests regardless these 48 errors and 2 failures?
Best regards, 
Ruslan

**Comment by pgagarinov** [comment]
I can see 2 different branches:

issue_12_RuslanM
issue_12_MalsagovR

Please remove one of them and then I'll take a look.

**Comment by ruslan-malsagov** [comment]
The branch issue_12_MalsagovR is now removed.
Thanks

**Comment by pgagarinov** [comment]
1) First of all you need to be able to reproduce each of the test failures from a command line.
Without that any debugging will be very painful. By reproducing a test I mean running
a single test without running the rest of the tests. The shorter your debug cycle (run->check->fix->run)
the more efficient you are. So please start with learning how to run each of the tests that failed.

A small hint - tests from SuiteRegression, SuiteCompare and SuiteSupport function are better
run via special scripts as opposed to using mlunitext.runtestcase just because it is easier.

For instance to reproduce a failure in

gras.ellapx.uncertcalc.test.regr.mlunit.SuiteRegression[demo3firstTest].testRegression

just run

gras.ellapx.uncertcalc.test.regr.run_regr_tests({'demo3firstTest'})

For other two test suites there are also special "run...tests" scripts that can accept configuraiton names
Please read a help header for more information run_regr_tests, but to keep it brief - "demo3firstTest"
is a configuration name that contains parameters of linear system so the command above
runs an ellipsoidal approximation calculation test for a specific system.

2) Based on what I see the tests fail because most of the code expects that the ellipsoidal approximation classes
you changed assume that the constructors accept "pDefObj,goodDirSetObj,timeLimsVec,relTol, absTol" i.e. they do not know
anything about the new "nTimePoints" parameter you added. I suggest you make this parameter a property that is not obligatory
so that a constructor can be called like this: 

TightEllApxBuilder(pDefObj,goodDirSetObj,timeLimsVec,relTol, absTol)

or

TightEllApxBuilder(pDefObj,goodDirSetObj,timeLimsVec,relTol, absTol,'nTimeGridPoints',nTimeGridPoints)

If 'nTimeGridPoints' property is not specified you assume that its value is taken from "N_TIME_POINTS" constant as before. 
Thus "N_TIME_POINTS" constant becomes a default value. This way the code that doesn't know about the new property (including the tests)
will work without any changes.

Please use modgen.common.parseparext for input argument parsing in the constructors 
(see https://github.com/SystemAnalysisDpt-CMC-MSU/ellipsoids/wiki/Coding-policy).

once you do 1 and 2 most if not all failures will go away and you can proceed with implementing the tests for the new functionality. The step #1 is a key here - it is just a waste of time to wait for the test results on the server to see if some single test is fixed, you just need to be able to run single tests from a command line manually, there is no way around it.

**Comment by ruslan-malsagov** [comment]
Good afternoon, Peter.

Could you please help? I used modgen.common.parseparext for input argument parsing as you recommended in your previous comment and I seem to know now how to reproduce tests from SuiteRegression, SuiteCompare and SuiteSupport functions. But I still can't reproduce two following fails:
> FAIL: elltool.reach.test.mlunit.ContinuousReachProjTestCase[demo3firstTest]/testSystem
> FAIL: elltool.reach.test.mlunit.ContinuousReachProjTestCase[demo3fourthTest]/testSystem.

I read the wiki about unit testing and tried to reproduce these failures by the command:
> mlunitext.runtestcase('elltool.reach.test.mlunit.ContinuousReachProjTestCase.testSystem').

However the output is the following: 
E
 INFO mlunitext.logprintf - ======================================================================
 INFO mlunitext.logprintf - ERROR: elltool.reach.test.mlunit.ContinuousReachProjTestCase/testSystem
 INFO mlunitext.logprintf - ----------------------------------------------------------------------
 INFO mlunitext.logprintf - Traceback (most recent call first): 
in C:\Users\Ruslan\Desktop\github\ellipsoids\products\+elltool\+reach\+test\+mlunit\AReachProjTestCase.m at line 48
in C:\Users\Ruslan\Desktop\github\ellipsoids\lib\mlunitext\+mlunitext\test_case.m at line 251
in C:\Users\Ruslan\Desktop\github\ellipsoids\lib\mlunitext\+mlunitext\test_suite.m at line 470
in C:\Users\Ruslan\Desktop\github\ellipsoids\lib\+modgen\+pcalc\auxdfevalpcomp.m at line 286
in C:\Users\Ruslan\Desktop\github\ellipsoids\lib\+modgen\+pcalc\auxdfeval.m at line 124
in C:\Users\Ruslan\Desktop\github\ellipsoids\lib\mlunitext\+mlunitext\+pcalc\auxdfeval.m at line 9
in C:\Users\Ruslan\Desktop\github\ellipsoids\lib\mlunitext\+mlunitext\test_suite.m at line 438
in C:\Users\Ruslan\Desktop\github\ellipsoids\lib\mlunitext\+mlunitext\text_test_runner.m at line 86
in C:\Users\Ruslan\Desktop\github\ellipsoids\lib\mlunitext\+mlunitext\runtestcase.m at line 150

Error: Not enough input arguments., Identifier: MATLAB:minrhs
 INFO mlunitext.logprintf - ----------------------------------------------------------------------
 INFO mlunitext.logprintf - Test case(s):
	elltool.reach.test.mlunit.ContinuousReachProjTestCase/testSystem,
 ran 1 test(s) in 0.031s(0.001m)
 INFO mlunitext.logprintf - FAILED (errors=1, failures=0)


ans = 

  text_test_result with properties:

        textOutFid: 1
    verbosityLevel: 1
     isDotDispMode: 1
        isAllShown: 0
            errors: {'elltool.reach.test.mlunit.ContinuousReachProjTestCase/testSystem'  [1x1950 char]}
          failures: {}
       should_stop: 0


+---------------------------------------------------------------------------------+
| << FAILED >> || TESTS: 1,  FAILURES: 0,  ERRORS: 1,  RUN TIME(sec.): 0.0035567  |
+---------------------------------------------------------------------------------+

So there are no two fails but there is one error (however there were no errors in the error_fail_list.txt which I got from the server). I also see "Error: Not enough input arguments" in the output, so I suppose this may be somehow caused by my insertions of an additional argument, but I'm not sure that I ran the corresponding tests correctly, therefore I still can't figure out how to fix this. Could you please help me how to reproduce these tests correctly?
Best regards,
Ruslan

**Comment by pgagarinov** [comment]
You need to provide additional input arguments either via "testParams" property in `mlunitext.runtestcase` or use `elltool.reach.test.run_cont_tests` function (recommended) like this:

elltool.reach.test.run_cont_tests('demo3firstTest','elltool.reach.test.mlunit.ContinuousReachProjTestCase','testSystem')

**Comment by ruslan-malsagov** [comment]
Thank you for the answer. I suppose then that in the "Example" section of the header of 
> products/+elltool/+reach/+test/run_cont_tests.m 

examples should be stated as `elltool.reach.test.run_cont_tests()` instead of  `elltool.reach.test.run_most_cont_tests()`, right? Should I fix this header as well?

Regards,
Ruslan

**Comment by pgagarinov** [comment]
Yes, please fix it.

**Comment by ruslan-malsagov** [comment]
It seems to me that these tests
> FAIL: elltool.reach.test.mlunit.ContinuousReachProjTestCase[demo3firstTest]/testSystem
> FAIL: elltool.reach.test.mlunit.ContinuousReachProjTestCase[demo3fourthTest]/testSystem

fail for any value of nTimePoints except 100. Is it just me or it is for real? Please, give me a hint in what direction I should be going, because I vainly tried to fix this and I really got stucked.

Also I wanted to ask if it is a better solution not to do 'nTimeGridPoints' a non-obligatory property, and just write in the constructor of
> ellipsoids/products/+gras/+ellapx/+lreachplain/ATightEllApxBuilder.m

something like this:
`import elltool.conf.Properties;`
`nTimeGridPoints=Properties.getNTimeGridPoints();`
`if isempty(nTimeGridPoints)
       nTimeGridPoints=ATightEllApxBuilder.N_TIME_POINTS; 
end`,
where `ATightEllApxBuilder.N_TIME_POINTS==100` (the default constant property value), so that we can pass the `nTimeGridPoints` value into a low-level EllTube builder class:
`self=self@gras.ellapx.gen.ATightEllApxBuilder(pDefObj,goodDirSetObj,timeLimsVec,nTimeGridPoints,relTol,absTol);`,
without changing the 'structure' of input arguments (because the constructor of low-level ET builder class initially accepts the nTimePoints argument).

**Comment by pgagarinov** [comment]
1) The  tests fail because you fixed a bug and now 'nTimeGridPoints'=200 by default, not 100 as before. 200 is used because this number is specified in ET configuration edited via elltool.editconf('default');
Changing the configuration is not an option because for production use we need 200 points but you need to make the tests keep using 100 points. One way to fix that is to pass nTimeGridPoints=100 into self.reachObfFactory.create call on line 87 in a parent class of ContinuousReachProjTestCase. This will cause self.reachObj having 100 time points as before and the tests should pass.

2) Please remove all calls without ";" at the end from ContinuousReachProjTestCase. These calls cause a lot of garbage console output. Someone obviously added it for debugging but one should have never committed it.

3) Please remove disp(nTimeGridPoints); call that you added - you can use such things for debugging but please never commit them. You can use log4 jogging as described in Wiki though to report a value of nTimeGridPoints in debug mode (logger.debug). what "debug" method of log4j logger does is controlled by log4jSettings section in xml configuration (use elltool.test.editconf('default') for editing).

4) Please always define constants for values like 100 in 

   function self=ATightIntEllApxBuilder(pDefObj,goodDirSetObj,...
                timeLimsVec,relTol,absTol,varargin)
            [~,~,sMethodName,nTimeGridPoints] = ...
                modgen.common.parseparext(varargin, ...
                {'selectionMethodForSMatrix','nTimeGridPoints';[],100},...

It is not ok to just specify 100 inline.

**Comment by ruslan-malsagov** [comment]
Good evening, Peter.

Thank you very much for help. Now I've done the 1)-4).

Regards,
Ruslan

**Comment by pgagarinov** [comment]
Yes, please

**Comment by ruslan-malsagov** [comment]
Peter, good afternoon.

In `ellipsoids/products/+elltool/+reach/+test/run_discr_tests.m` (https://github.com/SystemAnalysisDpt-CMC-MSU/ellipsoids/blob/master/products/%2Belltool/%2Breach/%2Btest/run_discr_tests.m) I found that for the `discrSecondTest` (in lines 65, 69, 73 and 77) `elltool.reach.test.mlunit.ContinuousReachTestCase` is loaded. Is it supposed to load Continuous instead of Discrete in run_discr_tests? 
When I change `elltool.reach.test.mlunit.ContinuousReachTestCase` to `elltool.reach.test.mlunit.DiscreteReachTestCase` I get an error saying `attempt to run test elltool.reach.test.mlunit.DiscreteReachTestCase[discrSecondTest_IsBackFalseIsEvolveFalse]/testGetCenter twice`. 

Regards,
Ruslan

**Comment by pgagarinov** [comment]
The tests in elltool.reach.test.mlunit.ContinuousReachTestCase are so well-written that they are universal for both discrete and continuous systems. What makes this possible is Dependency Injection pattern implemented in ContinuousReachTestCase and actually many other tests cases when the test cases are parameterized by factories. The factory decides which system to instantiate while tests deal with already created reachObj objects not knowing whether it is a discrete system or a continuous system. A type of system created by ReachFactory is determined by the sixth option input argument that is set to "true" in reach_discr_tests: "ReachFactory(confName, crm, crmSys, false, false, true)" and not set in run_cont_tests which corresponds to "false". So everything is correct, at least here but thanks to keeping your eyes open.

**Comment by ruslan-malsagov** [comment]
Thank you for response, got it.

**Comment by ruslan-malsagov** [comment]
Peter, good evening.

I made tests for nTimeGridPoints property for Continuous-systems case. Could you, please, see if it's alright? 

Regards,
Ruslan

**Comment by pgagarinov** [comment]
5) Calling Property.setNTimeGridPoints in a test is not safe unless you make sure that after the test the original value is restored. Please use set_up and tear_down (see https://github.com/SystemAnalysisDpt-CMC-MSU/ellipsoids/wiki/MLUNITEXT-unit-testing-framework-description) to restore the original value of the property so that other tests are not affected.

6) In the same test please consider a few additional cases:
   -  if you do call setNTimeGridPoints a number of time points used matches a number returned by getNTimeGridPoints 
   - if you pass nTimeGridPoints as a property into the constructor a number of points used matches nTimeGridPoints 

7) Add a test for the discrete system that checks that this property is ignored and number of time points depends on t0 and t1 only.

**Comment by ruslan-malsagov** [comment]
Thank you for your response. I've done 5)-7). Could you please take a look if everything is alright?

Regards,
Ruslan

**Comment by pgagarinov** [comment]
8) In 
products/+gras/+ellapx/+lreachuncert/MixedIntEllApxBuilder.m	

you parse nTimeGridPoints only to pass it further into 

products/+gras/+ellapx/+gen/ATightEllApxBuilder.m	

Please move parsing of nTimeGridProperty into +gen/ATightEllApxBuilder
and remove it from +lreachuncert/MixedIntEllApxBuilder.m	

Same goes for products/+gras/+ellapx/+lreachuncert/ExtIntEllApxBuilder.m	
and other descendants of gras.ellapx.gen.ATightEllApxBuilder.

Here is the rule - if you need to know the property directly in a descendant 
class - you parse it there, otherwise you pass it into a super-class for parsing.

[unparsedArgList,~,propA,probB]=modgen.parseparext(varargin,{'propA','probB'},[0 ?])
....

self=self@SuperClass(unparsedArgList{:});
self.propA=propA;
self.propB=propB;


9) In products/+gras/+ellapx/+lreachplain/ATightEllApxBuilder.m	

Get rid of "isempty" call in 

            if ~isempty(varargin)
                [~,~,nTimeGridPoints]=modgen.common.parseparext(varargin,...
                    {'nTimeGridPoints'},0,1);
            else
                nTimeGridPoints=ATightEllApxBuilder.N_TIME_POINTS;
            end
			
just by using default values of properties (see help header for parseparext)

10) products/+elltool/+reach/+test/+mlunit/ContinuousReachProjTestCase.m	

Please completely remove garbage calls, not just add ";" at the end.

11) My idea for "testSetMatchesGetNTimeGridPoints" test was not to call "setNTimeGridPoints" at all,
just check that if you do not do anything the number of points matches a number returned by 
Properties.getNTimeGridPoints();

**Comment by ruslan-malsagov** [comment]
Thank you for the explaination. I tried to work out steps 8)-11).

Regards,
Ruslan

**Comment by pgagarinov** [comment]
12) For parsing of nTimeGridPoints you use 'isObligatoryPropVec',false while this is not necessary - all properties with default values are not obligatory.

On the other hand, consider this:

>> [~,~,a]=modgen.common.parseparext({1,'alpha',2},{'alpha'})

a =

     2

>> [~,~,a]=modgen.common.parseparext({1},{'alpha'})

a =

     []

All this means that sMethod property that you parse in one of the descendant classes will be assigned an empty value if not specified. So you need to make it obligatory so that parseparext throws a human-readable error message. Otherwise the code will fail somewhere down the road with a mysterious error message. 

Once you are done with 12 and all the tests pass - please merge into master using the following algorithm:

1) Switch to master
2) Merge -> from -> origin/remotes/your branch (using local branch can dangerous because it can contain unpushed changes) 
3) Important !!! Set "no fast forward flag". Press OK
4) DO NOT PUSH JUST YET. Right click and open log. It should look like "petlya".
5) If eveything is ok - press push. 
6) Wait for the test results for updated master. If everything ok - switch to the next task which is #43

**Comment by ruslan-malsagov** [comment]
Thank you for your check-up. Could you please give me a hint: I got tests results and I got two errors in the following tests, which I can't reproduce on my computer:
`elltool.reach.test.mlunit.ContinuousReachTestNTimeGridPoints[demo3thirdTest_IsBackTrueIsEvolveFalse]/testSetMatchesGetNTimeGridPoints`
`elltool.reach.test.mlunit.ContinuousReachTestNTimeGridPoints[demo3thirdTest_IsBackTrueIsEvolveTrue]/testSetMatchesGetNTimeGridPoints`
They say:
`Error: internal approximation should be within external appproximation with tolerance 1.9999999999999999e-06, actual tolerance is 9.9814706473466686e-06, Identifier: GRAS:ELLAPX:SMARTDB:RELS:ELLTUBEBASIC:CHECKINTWITHINEXT:wrongInput:internalWithinExternal`.
I tried to change the NTimeGridPoints for different values in order to reproduce the errors, but I got tests passed every time. 

Regards,
Ruslan

**Comment by pgagarinov** [comment]
I suspect that setNTimeGridPoints is called prior to calling your tests in some other test. One of the scripts sets nTimeGridPoints to 250 which causes a problem. 

Here is what I found:

PropertiesTestCase.m	69	Properties.setNTimeGridPoints(timeGrid + 1);
Properties.m	201	function setNTimeGridPoints(nTimeGridPoints)
Properties.m	203	%   elltool.conf.Properties.setNTimeGridPoints(300);
anim6.m	13	Properties.setNTimeGridPoints(135);
bign.m	6	Properties.setNTimeGridPoints(135);
s_chapter05_section05_snippet03.m	1	elltool.conf.Properties.setNTimeGridPoints(250);
ContinuousReachTestNTimeGridPoints.m	26	Properties.setNTimeGridPoints(self.originalNTimeGridPoints);
ContinuousReachTestNTimeGridPoints.m	39	Properties.setNTimeGridPoints(N_TIME_POINTS);
DiscreteReachTestNTimeGridPoints.m	26	Properties.setNTimeGridPoints(self.originalNTimeGridPoints);
DiscreteReachTestNTimeGridPoints.m	39	Properties.setNTimeGridPoints(N_TIME_POINTS);
ell_demo3.m	17	Properties.setNTimeGridPoints(100);

Please find which tests cases call ell_demo3, anim6 and other scipts listed above. I know at least about

\products\+elltool\+demo\+test\+mlunit\TouchTestCase.m
\products\+elltool\+conf\+test\+mlunit\PropertiesTestCase.m

In these tests please add your set_up and tear_down methods, this should solve it.

**Comment by pgagarinov** [comment]
Do not forget about TouchTestCase - this tc calls demo scripts, at least s_chapter05_section05_snippet03 which sets nTimePoints=250 which seems to be a cause of the problem.

**Comment by ruslan-malsagov** [comment]
Peter, good afternoon.

Happy Defender of the Fatherland Day!

Thank you for your help and check-up. Sort of tricky problem. I've made corresponding insertions of set_up and tear_down functions.

Regards,
Ruslan

**Comment by ruslan-malsagov** [comment]
Peter, good evening.

I made all necessary changes in my branch, tests passed, and now I would like to merge into master using the following algorithm that you've mentioned before:

> 1) Switch to master
2) Merge -> from -> origin/remotes/your branch (using local branch can dangerous because it can contain unpushed changes) 
3) Important !!! Set "no fast forward flag". Press OK
4) DO NOT PUSH JUST YET. Right click and open log. It should look like "petlya".
5) If eveything is ok - press push. 

however I got stucked on the step 4, because when I opened log it didn't quite look like "petlya":
<img src="https://habrastorage.org/files/ec7/071/22f/ec707122fd304a16811d1cea1ee2b9aa.png"/>

I'm not sure that it's okay. I see green small "petlya" which comes from my branch and goes back. Actually I can hardly understand how it could happen. Could you please help me how to solve that occasion?

Regards,
Ruslan

**Comment by pgagarinov** [comment]
This doesn't look correct because you didn't check "no fast forward" flag when merging to master. You need to do check out a fresh version of master and do everything once again.

**Comment by ruslan-malsagov** [comment]
Well, actually I checked "no fast forward" flag, as you've mentioned in the step 3. Anyway, thank you, I'll try once again.

**Comment by ruslan-malsagov** [comment]
Could you please tell me what I supposed to do to "check out a fresh version of master"? Does this mean I should rebase my local master from the remotes/origin/master?

**Comment by pgagarinov** [comment]
This means that you need to do "switch/checkout" origin/master with "override if exists" flag. This will create a new local "master".

**Comment by ruslan-malsagov** [comment]
I got it, thank you very much. So now it looks alright, I suppose?
<img src="https://habrastorage.org/files/24e/70d/a07/24e70da0771846ff937ddf21c6a6b7db.png"/>

**Comment by pgagarinov** [comment]
Yes, it does. You can push. This is just a "razminochnaya" zadacha, you need to finish #43 to get "zachet" to please do not loose any time.

**Comment by ruslan-malsagov** [comment]
Thank you. I pushed. Could you please tell why it is said on Github that my branch is 1 commit behind master but when I press "compare" it says "nothing to compare"?

**Comment by pgagarinov** [comment]
Because master contains "merge" commit that your branch doesn't. This is normal, don't worry. Let us wait for the test results in master.

**Comment by ruslan-malsagov** [comment]
Okay, thanks the explaination.

**Comment by pgagarinov** [comment]
There still those two failures in master, not sure why. I'll fix this by myself. You - proceed to your new task.
